src/PostgREST/ApiRequest/Types.hs

```diff
@@ -92,6 +98,23 @@ type Cast = Text
 type Alias = Text
 type Hint = Text
 
+
+-- | A TypedField is a field with sufficient information to be read from JSON with `json_to_recordset`. In a future PR
+-- type fields will also be key for the data representations feature.
+-- Typed fields should have a nonempty field name and a nonempty type. We enforce this invariant in the `resolveField`
+-- function and keep the real constructor private. (It's not foolproof, you can still construct a `TypedField` using an
+-- existing TypedField and the Haskell `{ tfFieldName = "" }` syntax, so don't do that.)
```

Comment:
I think there are n ways to create an invalid field description like this, not just an empty string but invalid identifiers and so on. So there's no need to go to extremes with validation, which is why I didn't. At the same time a little validation can save some headache down the road, fail fast style.

Reply:
Actually I managed to do it while making this change with a typo in the SQL such that the field type ended up indeed empty! I can also imagine some request with `columns=,b` or whatever.

Anyhow I don't have a strong opinion either way. Happy to make the constructor public if that's preferred.

Reply:
Right, isn't that the idea with `TypedField` though, to be better validated and therefore less able to cause runtime errors down the line? The nature of bugs is that they seem impossible, otherwise you wouldn't have them!

Just making the point for the sake of rational debate, again, not married to having extra error checking here.

Reply:
I dropped the sanity checking from `TypedField`.